### DISCUSSION TITLE
Can old files in the Rotating files feature be named by date？

### DISCUSSION THREAD

**Discussion by jqfengVN1473** [gabime/spdlog#3159] [Q&A]
Currently, old files are suffixed with numbers, such as test.1.log test.2.log...
Now I want to use the current log as the extension for the old file，such as   test.2024-08-27.1.log , test.2024-08-27.2.log, What should I do?

**Comment by tt4g** [comment]
You can rename the log file with `file_event_handlers`.

https://github.com/gabime/spdlog/blob/a3a0c9d66386962fcaf0178fcae03ac77c1e0257/include/spdlog/common.h#L328C8-L339

Another way: create a custom sink: [Wiki: 4. Sinks (implementing your own sink)](https://github.com/gabime/spdlog/wiki/4.-Sinks#implementing-your-own-sink)

**Comment by jqfengVN1473** [comment]
Could you please give an example to illustrate？

**Reply by tt4g** [reply]
Example of renaming the previous day's rotated files after a day has passed (not tested):

```cpp
#include "spdlog/spdlog.h"
#include "spdlog/sinks/rotating_file_sink.h"
#include "spdlog/fmt/chrono.h"
#include "spdlog/details/os.h"

#include <cstdio>
#include <ctime>

class daily_rotate_handler
{
public:
    int max_files;
    spdlog::file_event_handlers file_event_handlers;

    explicit daily_rotate_handler(int max_files)
        : max_files(max_files),
          file_event_handlers(),
          rotation_tp_(next_rotation_tp_())
    {
        this->file_event_handlers.before_open = [this](spdlog::filename_t filename)
        {
            auto now = std::chrono::system_clock::now();
            if (!(now >= this->rotation_tp_)) {
                return;
            }
            auto prev_date = this->rotation_tp_ - std::chrono::hours(24);
            rotation_tp_ = next_rotation_tp_();

            // Rotation logic of rotating_file_sink:
            // https://github.com/gabime/spdlog/blob/a3a0c9d66386962fcaf0178fcae03ac77c1e0257/include/spdlog/sinks/rotating_file_sink-inl.h#L108-L129
            for (auto i = 0; i < this->max_files; ++i) {
                spdlog::filename_t src = spdlog::fmt_lib::format(
                    SPDLOG_FMT_STRING(
                        SPDLOG_FILENAME_T("logs/rotating.{}.txt")
                    ),
                    i
                );

                if (!spdlog::details::os::path_exists(src)) {
                    return;
                }

                spdlog::filename_t dest = spdlog::fmt_lib::format(
                    SPDLOG_FMT_STRING(
                        SPDLOG_FILENAME_T("logs/rotating.{:%Y-%m-%d}.{}.txt")
                    ),
                    prev_date,
                    i
                );
                spdlog::details::os::rename(src, dest);
            }
        };
    }

private:
    std::chrono::system_clock::time_point rotation_tp_;

    // Original daily_file_sink::next_rotation_tp_():
    // https://github.com/gabime/spdlog/blob/a3a0c9d66386962fcaf0178fcae03ac77c1e0257/include/spdlog/sinks/daily_file_sink.h#L149-L160
    std::chrono::system_clock::time_point next_rotation_tp_() {
        auto now = std::chrono::system_clock::now();

        std::time_t tnow = std::chrono::system_clock::to_time_t(now);
        std::tm date = spdlog::details::os::localtime(tnow);

        date.tm_hour = 0;
        date.tm_min = 0;
        date.tm_sec = 0;
        auto rotation_time = std::chrono::system_clock::from_time_t(std::mktime(&date));
        if (rotation_time > now) {
            return rotation_time;
        }
        return {rotation_time + std::chrono::hours(24)};
    }

};

int main()
{
    auto max_size = 1048576 * 5;
    auto rotate_on_open = false;

    daily_rotate_handler handler(3);

    auto logger = spdlog::rotating_logger_mt(
        "some_logger_name",
        "logs/rotating.txt",
        max_size,
        handler.max_files,
        rotate_on_open,
        handler.file_event_handlers
    );

    return 0;
}
```

**Reply by jqfengVN1473** [reply]
I am unable to compile your code even after copying it. Could you please provide a more complete version? I am not particularly familiar with spdlog, thank you

**Reply by tt4g** [reply]
My machine is currently not ready to build a C++ project.
Please fix the error yourself or provide a description of the error.

**Reply by tt4g** [reply]
I was able to use spdlog in Compiler Explorer (https://godbolt.org/z/zx41scx5f), so I fixed the code.
See [the original comments](https://github.com/gabime/spdlog/discussions/3159#discussioncomment-10470250).

**Reply by jqfengVN1473** [reply]
I used the code you provided and it compiled successfully, but the old file names are still suffixed with numbers, such as test.1.log

**Reply by tt4g** [reply]
If you want to do that, implement custom sink: [Wiki: 4. Sinks (implementing your own sink)](https://github.com/gabime/spdlog/wiki/4.-Sinks#implementing-your-own-sink)

**Reply by jqfengVN1473** [reply]
Could you please help me implement it? Thank you.

**Reply by jqfengVN1473** [reply]
I found that the java log library itself supports such a function. For example, the log files they generate are test.log, test-2024-01-23.1.log, test-2024-01-23.2.log, test-2024-01 -24.1.log, 2024-01-24.2.log

**Reply by tt4g** [reply]
> Could you please help me implement it? Thank you.

If you encounter problems implementing a custom sync, you can open an issue and get help.
However, the specification of the behavior of the date and size rotation is unclear.

Note a similar issue was created previously, but the specification of how it would work was unclear: https://github.com/gabime/spdlog/issues/2702#issuecomment-1501689716

> For example, the log files they generate are test.log, test-2024-01-23.1.log, test-2024-01-23.2.log, test-2024-01 -24.1.log, 2024-01-24.2.log

The above example does not show how the logs would be rotated.
For example, are you thinking about the following behavior specification?

* When is `test.log` rotated?
  * Suppose the 2024-01-23 log is written in `test.log`. When the date changes to 2024-01-24, will `test.log` be rotated to `test.log.2024-01-23.1.log`? Or does `test.log` remain as is?
    * If it doesn't rotate with the date, what would the file name be if it rotates on 2024-01-24? `test-2024-01-23.1.log`? Or `test-2024-01-24.1.log`?

Rotation of dates and sizes is often considered easy. However, the specifications must be carefully considered.
